1.

1.Simple linear regression uses one predictor, Multiple Linear Regression uses more than one.
2.Continuous allows smooth, nuanced changes, Indicator Variable is binary (0 or 1).
3.Adding an indicator allows different intercepts for each group, while keeping the same slope. Effect: The model captures group-based intercept shifts.
4.Interaction allows each group to have its own slope, capturing rate differences.
5.Indicator encoding (one-hot) represents non-binary categories.
Effect: The model captures fixed shifts between categories, not a continuous trend.

https://chatgpt.com/share/67369fce-e820-8001-9854-3a561b0145a6

2.

# Outcome and Predictor Variables
The outcome variable in this scenario is the **sales of sports equipment**, while the predictor variables are the **TV advertising budget (X1)** and **online advertising budget (X2)**. There may be an **interaction effect** between the two, meaning that the effectiveness of one advertising type could depend on the level of the other.

## Model without Interaction
Without interaction, the model is:
\[
Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + \epsilon
\]
Here, the effect of each budget is **independent**.

## Model with Interaction
With interaction, the model includes:
\[
Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + \beta_3 (X_1 \times X_2) + \epsilon
\]
In this model, \(\beta_3\) represents the **combined effect** of both budgets.

## Using the Models
Using the model **without interaction** gives predictions based on the independent contributions of each budget, while the model **with interaction** accounts for a potential **synergistic effect**, where high budgets in both TV and online ads may amplify sales.


https://chatgpt.com/share/6736a227-44ec-8001-9280-2f4df8e62055

3.

Data Preparation：Choose a couple of continuous, binary, and/or categorical variables as predictor variables. Ensure categorical variables are also converted to binary indicators if you're aiming for logistic regression.
Logistic Regression Model Specification:Use an additive model specification, i.e., a linear form where each predictor variable contributes to the outcome without interactions.Then, add interactions between predictor variables to see if there’s a synergistic effect.
Use the statsmodels.formula.api library (smf.logit) to fit logistic regression models for both additive and interaction-based models.
Since logistic regression uses log odds, interpret the coefficients as you would in a multivariate linear regression, focusing on whether predictors positively or negatively influence the outcome.
Use the .summary() table to evaluate statistical evidence (e.g., p-values) for each predictor.
Since direct logistic regression "best fit lines" aren’t straightforward, simulate it using the approach as if it were linear regression.
Generate random noise for the predictor values to simulate data points, then plot the model's best fit lines (one for additive, one for synergistic).


4.

In [2]:
import pandas as pd

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
# fail https://github.com/KeithGalli/pandas/blob/master/pokemon_data.csv
pokeaman = pd.read_csv(url) 


#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [3]:
import statsmodels.formula.api as smf

model1_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation) + Q("Sp. Def"):C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") * C(Generation)', data=pokeaman)

model2_fit = model2_spec.fit()
model2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     15.27
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           3.50e-27
Time:                        02:14:01   Log-Likelihood:                -3649.4
No. Observations:                 800   AIC:                             7323.
Df Residuals:                     788   BIC:                             7379.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          26.8971      5.246      5.127      0.000      16.599      37.195
C(Generation)[T.2]                 20.0449      7.821      2.563      0.011       4.692      35.398
C(Generation)[T.3]                 21.3662      6.998      3.053      0.002       7.629      35.103
C(Generation)[T.4]                 31.9575      8.235      3.881      0.000      15.793      48.122
C(Generation)[T.5]                  9.4926      7.883      1.204      0.229      -5.982      24.968
C(Generation)[T.6]                 22.2693      8.709      2.557      0.011       5.173      39.366
Q("Sp. Def")                        0.5634      0.071      7.906      0.000       0.423       0.703
Q("Sp. Def"):C(Generation)[T.2]    -0.2350      0.101     -2.316      0.021      -0.434      -0.036
Q("Sp. Def"):C(Generation)[T.3]    -0.3067      0.093     -3.300      0.001      -0.489      -0.124
Q("Sp. Def"):C(Generation)[T.4]    -0.3790      0.105     -3.600      0.000      -0.586      -0.172
Q("Sp. Def"):C(Generation)[T.5]    -0.0484      0.108     -0.447      0.655      -0.261       0.164
Q("Sp. Def"):C(Generation)[T.6]    -0.3083      0.112     -2.756      0.006      -0.528      -0.089
==============================================================================
Omnibus:                      337.229   Durbin-Watson:                   1.505
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2871.522
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                      11.649   Cond. No.                     1.40e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R-squared is 0.176 which means the explainabilitily is very low. It refers to the whole model's explainability. But the p value refes to one independent value's relation with dependent value. They are different. 

5.

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split

fifty_fifty_split_size = int(pokeaman.shape[0]*0.5)

# Replace "NaN" (in the "Type 2" column with "None")
pokeaman.fillna('None', inplace=True)

np.random.seed(130)
pokeaman_train,pokeaman_test = \
  train_test_split(pokeaman, train_size=fifty_fifty_split_size)
pokeaman_train


This code splits the pokeaman dataset into training and testing sets and prepares it for analysis:
Split Size Calculation: fifty_fifty_split_size is set to half the number of rows in pokeaman, which will be used as the training set size.
Handling Missing Values: The fillna function replaces any NaN values in the "Type 2" column with the string "None", ensuring there are no missing values in this column.
Random Seed: Setting a random seed (np.random.seed(130)) makes the data split reproducible.
Train-Test Split: train_test_split divides pokeaman into pokeaman_train and pokeaman_test, with the training set size specified by fifty_fifty_split_size.

In [ ]:
model_spec3 = smf.ols(formula='HP ~ Attack + Defense', 
                      data=pokeaman_train)
model3_fit = model_spec3.fit()
model3_fit.summary()

This code creates and fits a linear regression model to predict HP (Hit Points) based on the Attack and Defense attributes in the training dataset, pokeaman_train. It specifies HP as the dependent variable and Attack and Defense as independent variables. The model is then fit using .fit(), which estimates the effect of each predictor on HP. The .summary() function outputs detailed results, including how much of the variation in HP is explained by the model (R-squared), the coefficients, and the significance of each variable. This analysis helps in understanding how Attack and Defense contribute to predicting HP.

In [ ]:
yhat_model3 = model3_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model3_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model3)[0,1]**2)

This code calculates and compares the model’s R-squared values for both the training and test sets. First, it generates predicted HP values for the test set (yhat_model3) and retrieves the actual HP values (y). It then prints the "in-sample" R-squared, which shows how well the model fits the training data, and the "out-of-sample" R-squared, which is based on the correlation between predicted and actual test values. This comparison helps assess the model’s ability to generalize to new data.

In [ ]:
model4_linear_form = 'HP ~ Attack * Defense * Speed * Legendary'
model4_linear_form += ' * Q("Sp. Def") * Q("Sp. Atk")'
# DO NOT try adding '* C(Generation) * C(Q("Type 1")) * C(Q("Type 2"))'
# That's 6*18*19 = 6*18*19 possible interaction combinations...
# ...a huge number that will blow up your computer

model4_spec = smf.ols(formula=model4_linear_form, data=pokeaman_train)
model4_fit = model4_spec.fit()
model4_fit.summary()

This code defines and fits a complex linear regression model to predict HP (Hit Points) based on multiple features and their interactions. The model formula includes Attack, Defense, Speed, Legendary, Sp. Def, and Sp. Atk, with all possible interaction terms among them. Using Q() allows special character column names (like Sp. Def). The model excludes interactions with categorical variables Generation, Type 1, and Type 2 to prevent an excessively large number of combinations, which would make the model computationally unmanageable. After specifying the model, it is fit to the training data (pokeaman_train), and model4_fit.summary() provides details on the model’s performance, including R-squared values, coefficients, and significance levels of each predictor. This helps in understanding the complex relationships between HP and the selected features.

https://chatgpt.com/share/6736b392-c710-8001-8006-79e08ff6fabf

6.

In model4's design matrix, model4_spec.exog, adding numerous interaction terms significantly increases the variable count, leading to these issues:
Multicollinearity: While centering and scaling reduce some multicollinearity, the interaction terms remain highly correlated, resulting in an extremely high condition number (trillions), making the model unstable.
Generalization Problems: Multicollinearity causes overfitting to the training data, impairing the model’s ability to generalize to new data and leading to poor "out-of-sample" predictions.
Solutions: Reducing interaction terms, applying Principal Component Analysis (PCA), or using regularization methods (like Ridge Regression or Lasso) can lower multicollinearity and improve model stability and generalization.

https://chatgpt.com/share/6736b79c-eb60-8001-baa2-a466688b4558

7.

In [ ]:
# Here's something a little more reasonable...
model5_linear_form = 'HP ~ Attack + Defense + Speed + Legendary'
model5_linear_form += ' + Q("Sp. Def") + Q("Sp. Atk")'
model5_linear_form += ' + C(Generation) + C(Q("Type 1")) + C(Q("Type 2"))'

model5_spec = smf.ols(formula=model5_linear_form, data=pokeaman_train)
model5_fit = model5_spec.fit()
model5_fit.summary()

In [ ]:
yhat_model5 = model5_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model5_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model5)[0,1]**2)

In [ ]:
# Here's something a little more reasonable...
model6_linear_form = 'HP ~ Attack + Speed + Q("Sp. Def") + Q("Sp. Atk")'
# And here we'll add the significant indicators from the previous model
# https://chatgpt.com/share/81ab88df-4f07-49f9-a44a-de0cfd89c67c
model6_linear_form += ' + I(Q("Type 1")=="Normal")'
model6_linear_form += ' + I(Q("Type 1")=="Water")'
model6_linear_form += ' + I(Generation==2)'
model6_linear_form += ' + I(Generation==5)'

model6_spec = smf.ols(formula=model6_linear_form, data=pokeaman_train)
model6_fit = model6_spec.fit()
model6_fit.summary()

In [ ]:
yhat_model6 = model6_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model6_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2)

In [ ]:
# And here's a slight change that seems to perhaps improve prediction...
model7_linear_form = 'HP ~ Attack * Speed * Q("Sp. Def") * Q("Sp. Atk")'
model7_linear_form += ' + I(Q("Type 1")=="Normal")'
model7_linear_form += ' + I(Q("Type 1")=="Water")'
model7_linear_form += ' + I(Generation==2)'
model7_linear_form += ' + I(Generation==5)'

model7_spec = smf.ols(formula=model7_linear_form, data=pokeaman_train)
model7_fit = model7_spec.fit()
model7_fit.summary()

In [ ]:
yhat_model7 = model7_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model7_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2)

In [ ]:
# And here's a slight change that seems to perhas improve prediction...
model7_linear_form_CS = 'HP ~ scale(center(Attack)) * scale(center(Speed))'
model7_linear_form_CS += ' * scale(center(Q("Sp. Def"))) * scale(center(Q("Sp. Atk")))'
# We DO NOT center and scale indicator variables
model7_linear_form_CS += ' + I(Q("Type 1")=="Normal")'
model7_linear_form_CS += ' + I(Q("Type 1")=="Water")'
model7_linear_form_CS += ' + I(Generation==2)'
model7_linear_form_CS += ' + I(Generation==5)'

model7_CS_spec = smf.ols(formula=model7_linear_form_CS, data=pokeaman_train)
model7_CS_fit = model7_CS_spec.fit()
model7_CS_fit.summary().tables[-1] 
# "Cond. No." is NOW 15.4 due to centering and scaling

In [ ]:
# "Cond. No." WAS 2,340,000,000 WITHOUT to centering and scaling
model7_fit.summary().tables[-1]

I would like choose modle 6. 
Model 5: Includes basic predictors without interaction terms. It’s straightforward and interpretable, showing the direct effect of each variable on HP, but it doesn’t capture the relationships between predictors.
Model 6: Builds on Model 5, keeping only significant variables and indicator variables, resulting in a more streamlined model. It’s efficient in focusing on impactful predictors but still lacks interaction terms for complex relationships.
Model 7: Adds multiple interaction terms among the predictors, increasing the complexity. While this could improve predictive power, it introduced severe multicollinearity issues, making the model unstable.


https://chatgpt.com/share/6736c12d-7ecc-8001-87ec-17751d4ce13c

8.

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

# Placeholder for collecting performance metrics
in_sample_metrics = []
out_of_sample_metrics = []

# Simulate data and model performance metrics over multiple iterations
for i in range(100):
    # Simulate random data as "true" in-sample and out-of-sample values
    y_true_in_sample = np.random.rand(100)  # 100 random points as in-sample actual values
    y_true_out_of_sample = np.random.rand(100)  # 100 random points as out-of-sample actual values

    # Simulate model predictions with some added noise
    y_pred_in_sample = y_true_in_sample + np.random.normal(0, 0.1, 100)
    y_pred_out_of_sample = y_true_out_of_sample + np.random.normal(0, 0.1, 100)

    # Calculate a metric, e.g., Mean Squared Error (MSE)
    mse_in_sample = np.mean((y_true_in_sample - y_pred_in_sample) ** 2)
    mse_out_of_sample = np.mean((y_true_out_of_sample - y_pred_out_of_sample) ** 2)

    # Collect metrics
    in_sample_metrics.append(mse_in_sample)
    out_of_sample_metrics.append(mse_out_of_sample)

# Visualize the distribution of in-sample and out-of-sample metrics
plt.figure(figsize=(12, 6))
plt.hist(in_sample_metrics, bins=15, alpha=0.7, label="In-Sample MSE")
plt.hist(out_of_sample_metrics, bins=15, alpha=0.7, label="Out-of-Sample MSE")
plt.xlabel("Mean Squared Error")
plt.ylabel("Frequency")
plt.title("Distribution of In-Sample vs. Out-of-Sample Mean Squared Error")
plt.legend()
plt.show()


xplanation of Results and Purpose
Purpose: This code demonstrates the natural variation in model performance between in-sample and out-of-sample datasets. It simulates how a model might perform on data it has seen before (in-sample) compared to new data (out-of-sample).
In-Sample vs. Out-of-Sample Metrics: By generating metrics independently in each iteration, we observe potential differences in in-sample and out-of-sample errors. Often, in-sample metrics are slightly better because the model has been trained on this data, while out-of-sample metrics may be worse due to the model not having seen this data before.
Variability in Performance: Not using np.random.seed allows each iteration to generate different random values, which mimics real-world model training where performance can vary depending on the specific data samples. This helps in understanding the robustness of the model. If out-of-sample errors are consistently higher, this may indicate overfitting.
Practical Implication: By analyzing these metrics over multiple iterations, we gain insight into the stability and generalizability of the model. Ideally, the out-of-sample metrics should be close to the in-sample metrics, indicating good generalization.

https://chatgpt.com/share/6736c3a7-a5f4-8001-87d1-a9d17b2e9f18

9.

In [ ]:
model7_gen1_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model7_gen1_predict_future_fit = model7_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model7_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

This code calculates and compares the "in-sample" and "out-of-sample" \( R^2 \) values for two models. The first model (`model7_fit`) is trained on all data, while `model7_gen1_predict_future` is trained only on Generation 1 Pokémon data.

- The "in-sample" \( R^2 \) reflects each model's fit to its training data.
- The "out-of-sample" \( R^2 \) measures predictive accuracy on unseen data. For `model7_fit`, this is evaluated on the test dataset, and for `model7_gen1_predict_future_fit`, it’s calculated on data from Pokémon outside Generation 1.

This comparison helps determine how well a model trained on Generation 1 data alone generalizes to Pokémon from other generations.


In [ ]:
model7_gen1to5_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model7_gen1to5_predict_future_fit = model7_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model7_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

This code calculates and compares the "in-sample" and "out-of-sample" \( R^2 \) values for two models. The first model (`model7_fit`) is trained on all data, while `model7_gen1to5_predict_future` is trained on data from Generations 1 to 5 only. 

- The "in-sample" \( R^2 \) shows each model’s fit to its training data.
- The "out-of-sample" \( R^2 \) assesses prediction accuracy on unseen data: for `model7_fit`, this is the test dataset, and for `model7_gen1to5_predict_future_fit`, it’s Generation 6 Pokémon data.

This comparison helps evaluate how well a model trained without Generation 6 data generalizes to predicting Generation 6 Pokémon.


In [ ]:
model6_gen1_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model6_gen1_predict_future_fit = model6_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model6_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

This code calculates and compares the "in-sample" and "out-of-sample" \( R^2 \) values for two models. The first model (`model6_fit`) is trained on all data, while `model6_gen1_predict_future` is trained only on Generation 1 Pokémon data.

- The "in-sample" \( R^2 \) shows how well each model fits its respective training data.
- The "out-of-sample" \( R^2 \) assesses predictive performance on unseen data: for `model6_fit`, this is the test dataset; for `model6_gen1_predict_future_fit`, it’s data from Pokémon outside Generation 1.

This comparison helps evaluate the generalizability of a model trained only on Generation 1 data when predicting Pokémon from other generations.


In [ ]:
model6_gen1to5_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model6_gen1to5_predict_future_fit = model6_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model6_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

This code calculates and compares the "in-sample" and "out-of-sample" \( R^2 \) values for two models. The first model (`model6_fit`) is trained on all data, while `model6_gen1to5_predict_future` is trained on data from Generations 1 to 5 only.

- The "in-sample" \( R^2 \) shows each model's fit on its training data.
- The "out-of-sample" \( R^2 \) evaluates the models' predictive accuracy on unseen data. For `model6_fit`, the test set is used, while for `model6_gen1to5_predict_future_fit`, Generation 6 Pokémon data is used.

This comparison assesses how well a model trained on Generations 1 to 5 can generalize to predict Generation 6 Pokémon.


https://chatgpt.com/share/6736c56b-7470-8001-982c-49550afe4ca2